### ISSUE TITLE
paddle.grad for large models is very slow

### ISSUE THREAD

**Issue by holyseven** [PaddlePaddle/Paddle#33989]
- 版本、环境信息：
    1）PaddlePaddle版本：2.0.1
    2）CPU：请提供CPU型号，MKL/OpenBlas/MKLDNN/等数学库的使用情况
    3）GPU：CUDA 10.2.89, CUDNN7.6.5
    4）系统环境：Ubuntu18.04，Python 3.7.0
 
--- 

简单来说，用Vision Transformer的模型时，我想用`paddle.grad`计算输出对于输入图片的gradient，耗时25s左右；即使是计算 输出 对于 下述代码中`out`的 gradient，耗时依然在23-25s左右。
同样的代码，模型换成resnet50，耗时0.1s。

麻烦看一下是什么原因~

```python
import paddle
from interpretdl.data_processor.readers import preprocess_inputs
from datetime import datetime

img_path = 'assets/catdog.png'
# read the image and scale to [0, 1] with substract mean and divided by std.
imgs, data = preprocess_inputs(img_path, [224, 224, 3])  
print(imgs.shape, data.shape)  # (1, 224, 224, 3) (1, 3, 224, 224)

# from paddle.vision.models import resnet50
# paddle_model = resnet50(pretrained=True)

from assets.vision_transformer import ViT_small_patch16_224
# https://github.com/PaddlePaddle/PaddleClas/blob/release/2.2/ppcls/arch/backbone/model_zoo/vision_transformer.py
paddle_model = ViT_small_patch16_224()
MODEL_PATH = 'assets/ViT_small_patch16_224_pretrained.pdparams'
paddle_model.set_dict(paddle.load(MODEL_PATH))

data = paddle.to_tensor(data)
data.stop_gradient = False
out = paddle_model(data)
preds = paddle.argmax(out, axis=1)
label_onehot = paddle.nn.functional.one_hot(
                    paddle.to_tensor(preds), num_classes=out.shape[1])
target = paddle.sum(out * label_onehot, axis=1)

now = datetime.now()
g = paddle.grad(outputs=target, inputs=data)[0]
print(datetime.now() - now)  # 0:00:23.528224
```

**Comment by littletomatodonkey** [comment]
你好，可能有2个原因
1. vit的模型有278M，而resnet50只有98M，vit的网络结构比较零散，resnet50则比较整洁，所以vit要快一些
2. 在使用gpu的时候，前几次的耗时非常不准，gpu需要warmup，可以试着先跑10个iter，然后再统计后面10个iter的平均耗时，对比看下

**Comment by holyseven** [comment]
> 1. vit的模型有278M，而resnet50只有98M，vit的网络结构比较零散，resnet50则比较整洁，所以vit要快一些

嗯，好像的确不是vit的问题。我试了一下resnet101，resnet152，也很慢，几分钟了都没跑出来。小模型（包括mobilenet_v2, vgg13, vgg19）基本都是0.1s以内出结果。可以查一下具体是什么原因吗？

> 2. 在使用gpu的时候，前几次的耗时非常不准，gpu需要warmup，可以试着先跑10个iter，然后再统计后面10个iter的平均耗时，对比看下

时间没有变化，还是25s vs 0.1s。

除了上述问题，**即使是计算 输出 `target` 对于 下述代码中`out`的 gradient，耗时依然很长，这太不正常了。** `target`和`out`之间就一个加法的操作。

```python

import paddle
# from interpretdl.data_processor.readers import preprocess_inputs
from datetime import datetime
import numpy as np

img_path = 'assets/catdog.png'
# read the image and scale to [0, 1] with substract mean and divided by std.
data = np.random.rand(1, 3, 224, 224).astype(np.float32)
# imgs, data = preprocess_inputs(img_path, [224, 224, 3])  
print(data.shape, data.dtype)
# print(imgs.shape, data.shape)  # (1, 224, 224, 3) (1, 3, 224, 224)

from paddle.vision.models import resnet101, resnet50, mobilenet_v2, vgg13, vgg19
paddle_model = resnet101(pretrained=False)

# from assets.vision_transformer import ViT_small_patch16_224
# https://github.com/PaddlePaddle/PaddleClas/blob/release/2.2/ppcls/arch/backbone/model_zoo/vision_transformer.py
# paddle_model = ViT_small_patch16_224()
# MODEL_PATH = 'assets/ViT_small_patch16_224_pretrained.pdparams'
# paddle_model.set_dict(paddle.load(MODEL_PATH))

data = paddle.to_tensor(data)
data.stop_gradient = False
out = paddle_model(data)
preds = paddle.argmax(out, axis=1)
label_onehot = paddle.nn.functional.one_hot(
                    paddle.to_tensor(preds), num_classes=out.shape[1])
target = paddle.sum(out * label_onehot, axis=1)

now = datetime.now()
g = paddle.grad(outputs=target, inputs=out)[0]
print(datetime.now() - now)
```

相比torch，完全没有这个问题，以下为torch代码：
```python
# from interpretdl.data_processor.readers import preprocess_inputs
from datetime import datetime

# img_path = 'assets/catdog.png'
# read the image and scale to [0, 1] with substract mean and divided by std.
data = torch.rand((1, 3, 224, 224)).cuda()
print(imgs.shape, data.shape)  # (1, 224, 224, 3) (1, 3, 224, 224)

torch_model = models.resnet152().cuda()

# data = torch.to_tensor(data)
data.requires_grad = True
out = torch_model(data)
preds = torch.argmax(out, axis=1)
label_onehot = torch.nn.functional.one_hot(preds, num_classes=out.shape[1])
target = torch.sum(out * label_onehot, axis=1)

now = datetime.now()
from torch.autograd import grad
g = grad(outputs=target, inputs=data)[0]
print(datetime.now() - now) 
```

**Comment by littletomatodonkey** [comment]
你好，我复现了你的问题，目前在paddle.grad这块可以通过下面的方式去计算

```
# 这一步会重新构图，然后跑前向，再跑反向，拿到梯度信息，所以会慢一些
# g = paddle.grad(outputs=target, inputs=out)[0]
# 使用下面这种方式，直接基于已经构建好的图去跑反向
target.backward()
g = out.grad
```

另外， 关于paddle和torch的关于grad api的速度差异问题，我也反馈下，感谢你的反馈~

**Comment by holyseven** [comment]
上述代码很好地解决了耗时问题。虽然单张图片耗时上都是0.1s以内，但这样，想只计算中间层的gradient效率肯定是打折扣的。如果之后有反馈结果，也请告知，谢谢~
